x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java

```diff
@@ -262,4 +290,53 @@ public TimeValue visitSequenceParams(SequenceParamsContext ctx) {
                     text(numberCtx));
         }
     }
-}
+
+    private LogicalPlan pipe(PipeContext ctx, LogicalPlan plan) {
+        String name = text(ctx.IDENTIFIER());
+
+        if (SUPPORTED_PIPES.contains(name) == false) {
+            List<String> potentialMatches = StringUtils.findSimilar(name, SUPPORTED_PIPES);
+            
+            String msg = "Unrecognized pipe [{}]";
+            if (potentialMatches.isEmpty() == false) {
+                String matchString = potentialMatches.toString();
+                msg += ", did you mean " + (potentialMatches.size() == 1
+                        ? matchString
+                        : "any of " + matchString) + "?";
+            }
+            throw new ParsingException(source(ctx.IDENTIFIER()), msg, name);
+        }
+
+        switch (name) {
+            case "head":
+                Expression headLimit = pipeIntArgument(source(ctx), name, ctx.booleanExpression());
+                return new Head(source(ctx), headLimit, plan);
+            //new LimitWithOffset(source(ctx), headLimit, 0, plan)
+
+            case "tail":
+                Expression tailLimit = pipeIntArgument(source(ctx), name, ctx.booleanExpression());
+                // negate the limit
+                return new Tail(source(ctx), tailLimit, plan);
+            //return new LimitWithOffset(source(ctx), new Neg(tailLimit.source(), tailLimit), plan);
+
+            default:
+                throw new ParsingException(source(ctx), "Pipe [{}] is not supported yet", name);
+        }
+    }
+
+    private Expression pipeIntArgument(Source source, String pipeName, List<BooleanExpressionContext> exps) {
+        int size = CollectionUtils.isEmpty(exps) ? 0 : exps.size();
+        if (size != 1) {
+            throw new ParsingException(source, "Pipe [{}] expects exactly one argument but found [{}]", pipeName, size);
+        }
+        BooleanExpressionContext limitCtx = exps.get(0);
+        Expression expression = expression(limitCtx);
+
+        if (expression.dataType().isInteger() == false || expression.foldable() == false) {
+            throw new ParsingException(source(limitCtx), "Pipe [{}] expects a positive integer but found [{}]", pipeName, expression
```

Comment:
Where is the positive check taking place?

Reply:
There isn't any :) - the check was mainly to check the integer type. I've added one.